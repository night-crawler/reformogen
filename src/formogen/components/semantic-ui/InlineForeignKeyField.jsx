import React from 'react';
import { Form } from 'semantic-ui-react';
import Select from 'react-select';
import _ from 'lodash';
import PropTypes from 'prop-types';
import 'react-select/dist/react-select.css';

import Label from './Label';

import propTypes from '../fieldPropTypes';


makeOptions.propTypes = PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
}));

/**
 *
 * @param fieldData
 */
function makeOptions(fieldData) {
    return fieldData.map( ({ id, name }) => ({label: name, value: id}) );
}


InlineForeignKeyField.propTypes = propTypes;
export default function InlineForeignKeyField(props) {
    /*
        "name": "inspire_source",
        "verbose_name": "inspire source",
        "help_text": "",
        "blank": true,
        "null": false,
        "editable": true,
        "type": "ForeignKeyField",
        "required": false,
        "data": []
    */
    const handleChange = (val) => {
        props.onChange(null, {name: props.name, value: val ? val*1 : null});
    };

    let _props = {
        clearable: !props.required,
        closeOnSelect: true,
        disabled: !props.editable,
        multi: false,
        onChange: handleChange,
        options: makeOptions(props.data),
        placeholder: props.placeholder,
        simpleValue: true,
        value: props.value,
        inputProps: {type: 'react-type'},  // fixes broken semantic markup
        removeSelected: true,
        // rtl: this.state.rtl,
    };

    if (_.isFunction(props.updateProps)) {
        _props = props.updateProps(_props, props);
    }

    return (
        <Form.Field required={ props.required } disabled={ !props.editable } width={ props.layoutOpts.width }>
            <Label { ...props } />
            <Select { ..._props } />
            { !props.helpTextOnHover ? <span className="help-text">{ props.help_text }</span> : ''}
        </Form.Field>
    );
}
