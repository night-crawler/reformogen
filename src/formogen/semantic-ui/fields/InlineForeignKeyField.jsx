import React from 'react';
import PropTypes from 'prop-types';
import { Form } from 'semantic-ui-react';
import Select from 'react-select';
import { isEmpty } from 'lodash';

import { errorsType, layoutOptsType } from '../../fieldPropTypes';
import { FieldLabel } from '../FieldLabel';
import { ErrorsList } from '../ErrorsList';

import { extractIdentity, remapIdNameToLabelValue } from './utils';


InlineForeignKeyField.propTypes = {
  type: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.object]),
  data: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
  })),
  max_length: PropTypes.number,
  help_text: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  errors: errorsType,

  required: PropTypes.bool,
  editable: PropTypes.bool,

  helpTextOnHover: PropTypes.bool,
  layoutOpts: layoutOptsType,

  isRtl: PropTypes.bool,

  getOptionLabel: PropTypes.func,
  getOptionValue: PropTypes.func,

  onChange: PropTypes.func,
};
InlineForeignKeyField.defaultProps = {
  data: [],
  isRtl: false,
  getOptionLabel: ({ name }) => name,
  getOptionValue: ({ id }) => id,
};
export function InlineForeignKeyField(props) {
  const handleChange = ({ id }) => props.onChange(null, { 
    name: props.name, 
    value: id || null
  });

  return (
    <Form.Field
      required={ props.required }
      disabled={ !props.editable }
      width={ props.layoutOpts.width }
      error={ !isEmpty(props.errors) }
    >
      <FieldLabel { ...props } />
      <Select
        isClearable={ !props.required }
        closeOnSelect={ true }
        disabled={ !props.editable }
        isMulti={ false }
        onChange={ handleChange }
        options={ props.data }
        placeholder={ props.placeholder }
        simpleValue={ true }
        
        value={ props.data.filter(({ id }) => props.value === id ) }

        inputProps={ { type: 'react-type' } }
        removeSelected={ true }
        isRtl={ props.isRtl }

        getOptionLabel={ props.getOptionLabel }
        getOptionValue={ props.getOptionValue }
      />
      { !props.helpTextOnHover
        ? <span className='help-text'>{ props.help_text }</span>
        : ''
      }
      <ErrorsList messages={ props.errors } />
    </Form.Field>
  );
}
