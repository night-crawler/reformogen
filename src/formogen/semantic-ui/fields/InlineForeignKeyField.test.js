import { InlineForeignKeyField } from './InlineForeignKeyField';


it('<InlineForeignKeyField />', () => {
  const wrapper = shallow(
    <InlineForeignKeyField
      type='InlineForeignKeyField'
      name='InlineForeignKeyField: name'
      verbose_name='InlineForeignKeyField: verbose_name'
      help_text='InlineForeignKeyField: help_text'
      layoutOpts={ { width: 1 } }
    />
  );
  expect(wrapper).toMatchSnapshot();
});
