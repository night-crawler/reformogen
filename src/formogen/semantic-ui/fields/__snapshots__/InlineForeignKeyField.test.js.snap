// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InlineForeignKeyField /> 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InlineForeignKeyField
    data={Array []}
    getOptionLabel={[Function]}
    getOptionValue={[Function]}
    help_text="InlineForeignKeyField: help_text"
    isRtl={false}
    layoutOpts={
      Object {
        "width": 1,
      }
    }
    name="InlineForeignKeyField: name"
    type="InlineForeignKeyField"
    verbose_name="InlineForeignKeyField: verbose_name"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <FieldLabel
          data={Array []}
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          help_text="InlineForeignKeyField: help_text"
          isRtl={false}
          layoutOpts={
            Object {
              "width": 1,
            }
          }
          name="InlineForeignKeyField: name"
          type="InlineForeignKeyField"
          verbose_name="InlineForeignKeyField: verbose_name"
        />,
        <StateManager
          closeOnSelect={true}
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          disabled={true}
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          inputProps={
            Object {
              "type": "react-type",
            }
          }
          isClearable={true}
          isMulti={false}
          isRtl={false}
          onChange={[Function]}
          options={Array []}
          removeSelected={true}
          simpleValue={true}
          value={Array []}
        />,
        <span
          className="help-text"
        >
          InlineForeignKeyField: help_text
        </span>,
        <ErrorsList
          color="red"
          header="Errors"
        />,
      ],
      "disabled": true,
      "error": false,
      "required": undefined,
      "width": 1,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "data": Array [],
          "getOptionLabel": [Function],
          "getOptionValue": [Function],
          "help_text": "InlineForeignKeyField: help_text",
          "isRtl": false,
          "layoutOpts": Object {
            "width": 1,
          },
          "name": "InlineForeignKeyField: name",
          "type": "InlineForeignKeyField",
          "verbose_name": "InlineForeignKeyField: verbose_name",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "closeOnSelect": true,
          "defaultInputValue": "",
          "defaultMenuIsOpen": false,
          "defaultValue": null,
          "disabled": true,
          "getOptionLabel": [Function],
          "getOptionValue": [Function],
          "inputProps": Object {
            "type": "react-type",
          },
          "isClearable": true,
          "isMulti": false,
          "isRtl": false,
          "onChange": [Function],
          "options": Array [],
          "placeholder": undefined,
          "removeSelected": true,
          "simpleValue": true,
          "value": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "InlineForeignKeyField: help_text",
          "className": "help-text",
        },
        "ref": null,
        "rendered": "InlineForeignKeyField: help_text",
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "color": "red",
          "header": "Errors",
          "messages": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <FieldLabel
            data={Array []}
            getOptionLabel={[Function]}
            getOptionValue={[Function]}
            help_text="InlineForeignKeyField: help_text"
            isRtl={false}
            layoutOpts={
              Object {
                "width": 1,
              }
            }
            name="InlineForeignKeyField: name"
            type="InlineForeignKeyField"
            verbose_name="InlineForeignKeyField: verbose_name"
          />,
          <StateManager
            closeOnSelect={true}
            defaultInputValue=""
            defaultMenuIsOpen={false}
            defaultValue={null}
            disabled={true}
            getOptionLabel={[Function]}
            getOptionValue={[Function]}
            inputProps={
              Object {
                "type": "react-type",
              }
            }
            isClearable={true}
            isMulti={false}
            isRtl={false}
            onChange={[Function]}
            options={Array []}
            removeSelected={true}
            simpleValue={true}
            value={Array []}
          />,
          <span
            className="help-text"
          >
            InlineForeignKeyField: help_text
          </span>,
          <ErrorsList
            color="red"
            header="Errors"
          />,
        ],
        "disabled": true,
        "error": false,
        "required": undefined,
        "width": 1,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "data": Array [],
            "getOptionLabel": [Function],
            "getOptionValue": [Function],
            "help_text": "InlineForeignKeyField: help_text",
            "isRtl": false,
            "layoutOpts": Object {
              "width": 1,
            },
            "name": "InlineForeignKeyField: name",
            "type": "InlineForeignKeyField",
            "verbose_name": "InlineForeignKeyField: verbose_name",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "closeOnSelect": true,
            "defaultInputValue": "",
            "defaultMenuIsOpen": false,
            "defaultValue": null,
            "disabled": true,
            "getOptionLabel": [Function],
            "getOptionValue": [Function],
            "inputProps": Object {
              "type": "react-type",
            },
            "isClearable": true,
            "isMulti": false,
            "isRtl": false,
            "onChange": [Function],
            "options": Array [],
            "placeholder": undefined,
            "removeSelected": true,
            "simpleValue": true,
            "value": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "InlineForeignKeyField: help_text",
            "className": "help-text",
          },
          "ref": null,
          "rendered": "InlineForeignKeyField: help_text",
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "color": "red",
            "header": "Errors",
            "messages": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
